flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java

```diff
@@ -64,17 +67,35 @@
     public PulsarSourceEnumerator(
             PulsarSubscriber subscriber,
             StartCursor startCursor,
+            StopCursor stopCursor,
+            RangeGenerator rangeGenerator,
+            SourceConfiguration sourceConfiguration,
+            SplitEnumeratorContext<PulsarPartitionSplit> context) {
+        this(
+                subscriber,
+                startCursor,
+                stopCursor,
+                rangeGenerator,
+                sourceConfiguration,
+                context,
+                initialState());
+    }
+
+    public PulsarSourceEnumerator(
```

Comment:
I want to walk through the process of scale up/down to make sure my understanding is correct (especially in Shared subscription mode) : 

Whenever a scale up/down happen, the entire source will restart, first Enumerator starts, and each reader will be started and registerd to enumerator, and the addReader() method will be called. is this correct ?

In case of reader failure, the enumerator will not be restart, instead, addSplitsBack() will be called and then addReader() is called later.

Asking because I want to make sure whenever the scale up / down happens, the pendingSplits for that reader is in correct state (ensured by adding all fetched splits to each readers pending list at enumerator start all)

Reply:
The tests for shared subscription would be added to this PR soon. This would help you better understand the process.

Reply:
@imaffe The test has been added to this PR. You can check it.